### DISCUSSION TITLE
Window.postMessage() IPC

### DISCUSSION THREAD

**Discussion by saoirse-iontach** [os-js/OS.js#827] [Ideas]
Hello.

I was thinking about threading and debugger API...
In order to provide a debugger, we have to run in separate threads, like Worker or Iframe.

On an other hand, we could integrate page from many origin into a same window using iframe.
A corner case is using a file:// served distribution...

In those case, we should use some IPC mechanisms in order to communicate with a unique OSjs Core.
What we could do is to use Window.postMessage() to communicate between different instances,
say one top master OSjs Core and many framed slave OSjs cores.

This IPC could also be used to have many browser Windows for a unique remote desktop.
By example to allow multipes desktops for a unique session.

In order to properly work, we'll need to provide a generic mechanism to proxify API and services...
(by example, OSjs.make('osjs/vfs') in a slave should provide an IPC proxy to the top master Core)
This also require some configuration and security (especially with file:// origin).

**Comment by andersevenrud** [comment]
The core event bus actually hooks into the window messages, and iframe apps can use this to provide an IPC mechanism.

This is very basic and only provides an interface with custom events.

It would be cool to have a slave core instance as suggested here.

Providing services over a mechanism like this is fairly trivial, but handling shared states (like sharing a session) will require some work.

**Reply by saoirse-iontach** [reply]
I can't find where are those hooks. Only in v2 onPostMessage.

**Reply by andersevenrud** [reply]
https://github.com/os-js/osjs-client/blob/f3e5c1323197eae19a8dc4f337c234d6aa301113/src/core.js#L445

https://manual.os-js.org/tutorial/iframe/
